### DISCUSSION TITLE
Is it a must that notify is the final command in a pipe of several commands?

### DISCUSSION THREAD

**Discussion by TheeEclipse** [projectdiscovery/notify#441] [Q&A]
I am trying to pipe my workflow like:
```cat domains.txt|subfinder|notify   -provider-config subs.yaml | custom-bugscans ```

Here the idea is I get  subdomains and notify to my subdomains BOT then take the subdomains and pipe to custom scanner and pipe those to my BUGS BOT.

But it doesnt work:

sample:
```echo test  |notify  -provider-config  /root/httpx.yaml  -silent | xargs -I {} echo ONE.{}```

OUTPUT:
```
Intigriti:~# echo test  |notify  -provider-config  /root/httpx.yaml  -silent | xargs -I {} echo ONE.{}
ONE.test
Intigriti:~#
```

The final echo works as you can see BUT ❌NO NOTIFY NOTIFICATION WAS SENT !!❌

As long as notify is not the last in a pipe, it doesnt work at all

Am I missing something?

**Comment by TheeEclipse** [comment]
Got a way to this.

**Comment by bad-antics** [comment]
You're right that notify has buffering behavior. Here's what's happening and how to fix it:

**The Issue:**
Notify buffers input for efficiency, and when piped further, the output buffering can interfere with notification timing.

**Solution 1: Use `tee` to split the stream**
```bash
echo test | tee >(notify -provider-config /root/httpx.yaml -silent) | xargs -I {} echo ONE.{}
```

**Solution 2: Use `-bulk` flag**
```bash
echo test | notify -provider-config /root/httpx.yaml -silent -bulk | xargs -I {} echo ONE.{}
```

**Solution 3: Force immediate output**
```bash
echo test | stdbuf -oL notify -provider-config /root/httpx.yaml -silent | xargs -I {} echo ONE.{}
```

**Best approach for your workflow:**
```bash
cat domains.txt | subfinder | tee >(notify -provider-config subs.yaml -silent) | custom-bugscans | notify -provider-config bugs.yaml -silent
```

**Alternative - Use file-based approach:**
```bash
subfinder -dL domains.txt -o subs.txt && notify -data subs.txt -provider-config subs.yaml
cat subs.txt | custom-bugscans | notify -provider-config bugs.yaml
```

The `tee` approach is cleanest - it sends data to both notify AND the next command in parallel.

**Why it happens:**
Notify reads stdin fully before sending to avoid partial messages. With tee, it gets a copy while the original flows through.

**Comment by bad-antics** [comment]
You're right that notify has buffering behavior. Here's what's happening and how to fix it:

**The Issue:**
Notify buffers input for efficiency, and when piped further, the output buffering can interfere with notification timing.

**Solution 1: Use `tee` to split the stream**
```bash
echo test | tee >(notify -provider-config /root/httpx.yaml -silent) | xargs -I {} echo ONE.{}
```

**Solution 2: Use `-bulk` flag**
```bash
echo test | notify -provider-config /root/httpx.yaml -silent -bulk | xargs -I {} echo ONE.{}
```

**Solution 3: Force immediate output**
```bash
echo test | stdbuf -oL notify -provider-config /root/httpx.yaml -silent | xargs -I {} echo ONE.{}
```

**Best approach for your workflow:**
```bash
cat domains.txt | subfinder | tee >(notify -provider-config subs.yaml -silent) | custom-bugscans | notify -provider-config bugs.yaml -silent
```

**Alternative - Use file-based approach:**
```bash
subfinder -dL domains.txt -o subs.txt && notify -data subs.txt -provider-config subs.yaml
cat subs.txt | custom-bugscans | notify -provider-config bugs.yaml
```

The `tee` approach is cleanest - it sends data to both notify AND the next command in parallel.

**Why it happens:**
Notify reads stdin fully before sending to avoid partial messages. With tee, it gets a copy while the original flows through.

**Comment by bad-antics** [comment]
No, notify doesn't have to be the last command — it **does** pass through stdin to stdout by default. But there's a timing/buffering issue.

**The fix — use `-silent`:**
```bash
cat domains.txt | subfinder -silent | notify -provider-config subs.yaml -silent | custom-bugscans
```

**Why it wasn't working:**

1. notify's **progress/status output** was mixing with the data stream. `-silent` suppresses everything except the actual data passthrough.

2. **Buffering** — notify processes and sends notifications asynchronously. The pipe might close before notify finishes sending.

**Alternative approach — use `tee` to split the stream:**
```bash
cat domains.txt | subfinder -silent | tee >(notify -provider-config subs.yaml -silent) | custom-bugscans
```

This sends the subdomains to both notify AND your custom scanner simultaneously using process substitution.

**Or use a file-based workflow for reliability:**
```bash
# Step 1: Get subdomains and save
subfinder -dL domains.txt -silent -o subs.txt

# Step 2: Notify about new subs
notify -data subs.txt -provider-config subs.yaml -silent

# Step 3: Scan
cat subs.txt | custom-bugscans | notify -provider-config bugs.yaml -silent
```

The pipe approach works, but you need `-silent` on both subfinder and notify to keep the data stream clean.
